Testing issues on `main` related to monitoring tests

Reproducible failures on `main`:

https://github.com/numba/numba/actions/runs/24277927946/job/71254852424

```
======================================================================
ERROR: test_raising_callback_unwinds_from_jit_on_unwind_path (numba.tests.test_sys_monitoring.TestMonitoring.test_raising_callback_unwinds_from_jit_on_unwind_path)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.14.3/x64/lib/python3.14/site-packages/numba/tests/test_sys_monitoring.py", line 723, in test_raising_callback_unwinds_from_jit_on_unwind_path
    raising()
    ~~~~~~~^^
  File "/opt/hostedtoolcache/Python/3.14.3/x64/lib/python3.14/site-packages/numba/tests/test_sys_monitoring.py", line 714, in raising
    raise LocalException(msg_execution)
numba.tests.test_sys_monitoring.TestMonitoring.test_raising_callback_unwinds_from_jit_on_unwind_path.<locals>.LocalException: deliberately broken execution

======================================================================
FAIL: test_profiler (numba.tests.test_profiler.TestProfiler.test_profiler)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.14.3/x64/lib/python3.14/site-packages/numba/tests/test_profiler.py", line 91, in test_profiler
    self.check_profiler_dot(dot, dot)
    ~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.14.3/x64/lib/python3.14/site-packages/numba/tests/test_profiler.py", line 83, in check_profiler_dot
    check_stats_for_key(stats, cfunc.py_func.__code__, n_calls)
    ~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.14.3/x64/lib/python3.14/site-packages/numba/tests/test_profiler.py", line 77, in check_stats_for_key
    self.assertIn(expected_key, stats.stats)
    ~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: ('test_profiler.py', 16, 'dot') not found in {('common.py', 49, '__init__'): (1, 1, 1.1979e-05, 2.4183000000000002e-05, {('npytypes.py', 444, '__init__'): (1, 1, 1.1979e-05, 2.4183000000000002e-05)}), ('npytypes.py', 444, '__init__'): (1, 1, 1.0772000000000001e-05, 4.4764e-05, {('abstract.py', 61, '__call__'): (1, 1, 1.0772000000000001e-05, 4.4764e-05)}), ('numpy_support.py', 85, 'from_dtype'): (1, 1, 3.3710000000000004e-06, 4.6230000000000005e-06, {('typeof.py', 246, '_typeof_ndarray'): (1, 1, 3.3710000000000004e-06, 4.6230000000000005e-06)}), ('typeof.py', 246, '_typeof_ndarray'): (1, 1, 1.8153e-05, 0.00010337100000000001, {('functools.py', 978, 'wrapper'): (1, 1, 1.8153e-05, 0.00010337100000000001)}), ('~', 0, "<method 'disable' of '_lsprof.Profiler' objects>"): (1, 1, 3.0497000000000003e-05, 3.0497000000000003e-05, {}), ('dispatcher.py', 678, 'typeof_pyval'): (1, 1, 7.114000000000001e-06, 0.000142852, {}), ('numpy_support.py', 221, 'map_layout'): (1, 1, 4.227e-06, 4.227e-06, {('typeof.py', 246, '_typeof_ndarray'): (1, 1, 4.227e-06, 4.227e-06)}), ('<string>', 1, '<lambda>'): (1, 1, 6.155e-06, 6.999e-06, {('typeof.py', 27, 'typeof'): (1, 1, 6.155e-06, 6.999e-06)}), ('npytypes.py', 482, 'key'): (4, 4, 2.79e-06, 2.79e-06, {('abstract.py', 124, '__eq__'): (2, 2, 1.374e-06, 1.374e-06), ('abstract.py', 121, '__hash__'): (2, 2, 1.416e-06, 1.416e-06)}), ('abstract.py', 61, '__call__'): (1, 1, 1.076e-05, 7.5834e-05, {('typeof.py', 246, '_typeof_ndarray'): (1, 1, 1.076e-05, 7.5834e-05)}), ('misc.py', 63, 'unliteral'): (1, 1, 2.4050000000000003e-06, 3.3760000000000004e-06, {('common.py', 49, '__init__'): (1, 1, 2.4050000000000003e-06, 3.3760000000000004e-06)}), ('typeof.py', 27, 'typeof'): (1, 1, 5.7290000000000005e-06, 0.00013033600000000002, {('dispatcher.py', 678, 'typeof_pyval'): (1, 1, 5.7290000000000005e-06, 0.00013033600000000002)}), ('abstract.py', 124, '__eq__'): (1, 1, 2.105e-06, 3.479e-06, {('~', 0, "<method 'get' of 'dict' objects>"): (1, 1, 2.105e-06, 3.479e-06)}), ('abstract.py', 121, '__hash__'): (2, 4, 6.7650000000000005e-06, 1.1842000000000001e-05, {('~', 0, "<method 'get' of 'dict' objects>"): (1, 1, 3.428e-06, 7.865e-06), ('~', 0, "<method 'add' of 'set' objects>"): (1, 1, 1.074e-06, 3.977e-06), ('~', 0, '<built-in method builtins.hash>'): (2, 2, 2.263e-06, 3.7370000000000004e-06)}), ('abstract.py', 93, '__init__'): (1, 1, 7.36e-07, 7.36e-07, {('common.py', 49, '__init__'): (1, 1, 7.36e-07, 7.36e-07)}), ('abstract.py', 96, 'key'): (2, 2, 6.480000000000001e-07, 6.480000000000001e-07, {('abstract.py', 121, '__hash__'): (2, 2, 6.480000000000001e-07, 6.480000000000001e-07)}), ('abstract.py', 118, '__str__'): (1, 1, 6.210000000000001e-07, 6.210000000000001e-07, {('npytypes.py', 444, '__init__'): (1, 1, 6.210000000000001e-07, 6.210000000000001e-07)}), ('abstract.py', 49, '_intern'): (1, 1, 4.575e-06, 2.031e-05, {('abstract.py', 61, '__call__'): (1, 1, 4.575e-06, 2.031e-05)}), ('weakref.py', 325, '__getitem__'): (1, 1, 2.662e-06, 2.662e-06, {('functools.py', 893, 'dispatch'): (1, 1, 2.662e-06, 2.662e-06)}), ('<frozen abc>', 117, '__instancecheck__'): (3, 3, 3.2900000000000003e-06, 7.894000000000001e-06, {('~', 0, '<built-in method builtins.isinstance>'): (3, 3, 3.2900000000000003e-06, 7.894000000000001e-06)}), ('<frozen importlib._bootstrap>', 648, 'parent'): (1, 1, 4.019e-06, 4.9950000000000005e-06, {('common.py', 49, '__init__'): (1, 1, 4.019e-06, 4.9950000000000005e-06)}), ('functools.py', 893, 'dispatch'): (1, 1, 5.206000000000001e-06, 8.28e-06, {('functools.py', 978, 'wrapper'): (1, 1, 5.206000000000001e-06, 8.28e-06)}), ('functools.py', 978, 'wrapper'): (1, 1, 5.957000000000001e-06, 0.000117608, {('typeof.py', 27, 'typeof'): (1, 1, 5.957000000000001e-06, 0.000117608)}), ('~', 0, "<method 'get' of 'dict' objects>"): (1, 1, 4.391e-06, 1.5735e-05, {('abstract.py', 49, '_intern'): (1, 1, 4.391e-06, 1.5735e-05)}), ('~', 0, "<method 'add' of 'set' objects>"): (1, 1, 1.4250000000000001e-06, 5.402e-06, {('dispatcher.py', 678, 'typeof_pyval'): (1, 1, 1.4250000000000001e-06, 5.402e-06)}), ('~', 0, '<built-in method __new__ of type object at 0x7fa76cb40940>'): (1, 1, 8.440000000000001e-07, 8.440000000000001e-07, {('<string>', 1, '<lambda>'): (1, 1, 8.440000000000001e-07, 8.440000000000001e-07)}), ('~', 0, "<method 'rpartition' of 'str' objects>"): (1, 1, 9.76e-07, 9.76e-07, {('<frozen importlib._bootstrap>', 648, 'parent'): (1, 1, 9.76e-07, 9.76e-07)}), ('~', 0, '<built-in method builtins.getattr>'): (2, 2, 1.6680000000000002e-06, 1.6680000000000002e-06, {('numpy_support.py', 85, 'from_dtype'): (1, 1, 1.252e-06, 1.252e-06), ('misc.py', 63, 'unliteral'): (1, 1, 4.16e-07, 4.16e-07)}), ('~', 0, '<built-in method builtins.hasattr>'): (1, 1, 5.550000000000001e-07, 5.550000000000001e-07, {('misc.py', 63, 'unliteral'): (1, 1, 5.550000000000001e-07, 5.550000000000001e-07)}), ('~', 0, '<built-in method builtins.hash>'): (2, 4, 3.013e-06, 5.924e-06, {('abstract.py', 121, '__hash__'): (4, 2, 3.013e-06, 5.924e-06)}), ('~', 0, '<built-in method builtins.isinstance>'): (4, 4, 4.925e-06, 1.2819e-05, {('common.py', 49, '__init__'): (1, 1, 8.310000000000001e-07, 3.097e-06), ('npytypes.py', 444, '__init__'): (2, 2, 3.5600000000000002e-06, 9.188e-06), ('typeof.py', 246, '_typeof_ndarray'): (1, 1, 5.34e-07, 5.34e-07)}), ('~', 0, '<built-in method _abc.get_cache_token>'): (1, 1, 4.1200000000000004e-07, 4.1200000000000004e-07, {('functools.py', 893, 'dispatch'): (1, 1, 4.1200000000000004e-07, 4.1200000000000004e-07)}), ('~', 0, '<built-in method _abc._abc_instancecheck>'): (3, 3, 4.604e-06, 4.604e-06, {('<frozen abc>', 117, '__instancecheck__'): (3, 3, 4.604e-06, 4.604e-06)})}

======================================================================
FAIL: test_profiler_for_raising_function (numba.tests.test_profiler.TestProfiler.test_profiler_for_raising_function)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.14.3/x64/lib/python3.14/site-packages/numba/tests/test_profiler.py", line 95, in test_profiler_for_raising_function
    self.check_profiler_dot(call_raising_dot, raising_dot)
    ~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.14.3/x64/lib/python3.14/site-packages/numba/tests/test_profiler.py", line 83, in check_profiler_dot
    check_stats_for_key(stats, cfunc.py_func.__code__, n_calls)
    ~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.14.3/x64/lib/python3.14/site-packages/numba/tests/test_profiler.py", line 77, in check_stats_for_key
    self.assertIn(expected_key, stats.stats)
    ~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: ('test_profiler.py', 29, 'raising_dot') not found in {('common.py', 49, '__init__'): (1, 1, 1.2780000000000001e-05, 2.4871000000000003e-05, {('npytypes.py', 444, '__init__'): (1, 1, 1.2780000000000001e-05, 2.4871000000000003e-05)}), ('npytypes.py', 444, '__init__'): (1, 1, 9.85e-06, 4.4368e-05, {('abstract.py', 61, '__call__'): (1, 1, 9.85e-06, 4.4368e-05)}), ('numpy_support.py', 85, 'from_dtype'): (1, 1, 3.766e-06, 4.800000000000001e-06, {('typeof.py', 246, '_typeof_ndarray'): (1, 1, 3.766e-06, 4.800000000000001e-06)}), ('typeof.py', 246, '_typeof_ndarray'): (1, 1, 1.8057e-05, 0.000101811, {('functools.py', 978, 'wrapper'): (1, 1, 1.8057e-05, 0.000101811)}), ('test_profiler.py', 39, 'call_raising_dot'): (123, 123, 0.00049909, 0.0007079780000000001, {}), ('serialize.py', 30, '_numba_unpickle'): (123, 123, 5.2537e-05, 6.541e-05, {('test_profiler.py', 39, 'call_raising_dot'): (123, 123, 5.2537e-05, 6.541e-05)}), ('dispatcher.py', 678, 'typeof_pyval'): (1, 1, 8.564000000000001e-06, 0.000143478, {('test_profiler.py', 39, 'call_raising_dot'): (1, 1, 8.564000000000001e-06, 0.000143478)}), ('numpy_support.py', 221, 'map_layout'): (1, 1, 2.9980000000000003e-06, 2.9980000000000003e-06, {('typeof.py', 246, '_typeof_ndarray'): (1, 1, 2.9980000000000003e-06, 2.9980000000000003e-06)}), ('<string>', 1, '<lambda>'): (1, 1, 6.813000000000001e-06, 7.723e-06, {('typeof.py', 27, 'typeof'): (1, 1, 6.813000000000001e-06, 7.723e-06)}), ('npytypes.py', 482, 'key'): (4, 4, 2.6760000000000003e-06, 2.6760000000000003e-06, {('abstract.py', 124, '__eq__'): (2, 2, 1.2870000000000001e-06, 1.2870000000000001e-06), ('abstract.py', 121, '__hash__'): (2, 2, 1.3890000000000002e-06, 1.3890000000000002e-06)}), ('abstract.py', 61, '__call__'): (1, 1, 9.55e-06, 7.539100000000001e-05, {('typeof.py', 246, '_typeof_ndarray'): (1, 1, 9.55e-06, 7.539100000000001e-05)}), ('misc.py', 63, 'unliteral'): (1, 1, 2.921e-06, 3.734e-06, {('common.py', 49, '__init__'): (1, 1, 2.921e-06, 3.734e-06)}), ('typeof.py', 27, 'typeof'): (1, 1, 6.6020000000000005e-06, 0.000129899, {('dispatcher.py', 678, 'typeof_pyval'): (1, 1, 6.6020000000000005e-06, 0.000129899)}), ('abstract.py', 124, '__eq__'): (1, 1, 2.313e-06, 3.6000000000000003e-06, {('~', 0, "<method 'get' of 'dict' objects>"): (1, 1, 2.313e-06, 3.6000000000000003e-06)}), ('abstract.py', 121, '__hash__'): (2, 4, 6.992e-06, 1.179e-05, {('~', 0, "<method 'get' of 'dict' objects>"): (1, 1, 3.0360000000000002e-06, 7.791e-06), ('~', 0, "<method 'add' of 'set' objects>"): (1, 1, 1.1100000000000002e-06, 3.999e-06), ('~', 0, '<built-in method builtins.hash>'): (2, 2, 2.8460000000000004e-06, 4.063e-06)}), ('abstract.py', 93, '__init__'): (1, 1, 4.16e-07, 4.16e-07, {('common.py', 49, '__init__'): (1, 1, 4.16e-07, 4.16e-07)}), ('abstract.py', 96, 'key'): (2, 2, 6.350000000000001e-07, 6.350000000000001e-07, {('abstract.py', 121, '__hash__'): (2, 2, 6.350000000000001e-07, 6.350000000000001e-07)}), ('abstract.py', 118, '__str__'): (1, 1, 3.87e-07, 3.87e-07, {('npytypes.py', 444, '__init__'): (1, 1, 3.87e-07, 3.87e-07)}), ('abstract.py', 49, '_intern'): (1, 1, 5.037e-06, 2.1473000000000003e-05, {('abstract.py', 61, '__call__'): (1, 1, 5.037e-06, 2.1473000000000003e-05)}), ('~', 0, "<method 'disable' of '_lsprof.Profiler' objects>"): (1, 1, 3.8092e-05, 3.8092e-05, {}), ('~', 0, '<built-in method _pickle.loads>'): (1, 1, 1.2873e-05, 1.2873e-05, {('serialize.py', 30, '_numba_unpickle'): (1, 1, 1.2873e-05, 1.2873e-05)}), ('weakref.py', 325, '__getitem__'): (1, 1, 2.465e-06, 2.465e-06, {('functools.py', 893, 'dispatch'): (1, 1, 2.465e-06, 2.465e-06)}), ('<frozen abc>', 117, '__instancecheck__'): (3, 3, 2.674e-06, 7.5770000000000005e-06, {('~', 0, '<built-in method builtins.isinstance>'): (3, 3, 2.674e-06, 7.5770000000000005e-06)}), ('<frozen importlib._bootstrap>', 648, 'parent'): (1, 1, 4.247e-06, 5.275e-06, {('common.py', 49, '__init__'): (1, 1, 4.247e-06, 5.275e-06)}), ('functools.py', 893, 'dispatch'): (1, 1, 4.728e-06, 7.775000000000001e-06, {('functools.py', 978, 'wrapper'): (1, 1, 4.728e-06, 7.775000000000001e-06)}), ('functools.py', 978, 'wrapper'): (1, 1, 5.988e-06, 0.000115574, {('typeof.py', 27, 'typeof'): (1, 1, 5.988e-06, 0.000115574)}), ('~', 0, "<method 'get' of 'dict' objects>"): (1, 1, 5.045e-06, 1.6436e-05, {('abstract.py', 49, '_intern'): (1, 1, 5.045e-06, 1.6436e-05)}), ('~', 0, "<method 'add' of 'set' objects>"): (1, 1, 1.016e-06, 5.015e-06, {('dispatcher.py', 678, 'typeof_pyval'): (1, 1, 1.016e-06, 5.015e-06)}), ('~', 0, '<built-in method __new__ of type object at 0x7fd5f8140940>'): (1, 1, 9.100000000000001e-07, 9.100000000000001e-07, {('<string>', 1, '<lambda>'): (1, 1, 9.100000000000001e-07, 9.100000000000001e-07)}), ('~', 0, "<method 'rpartition' of 'str' objects>"): (1, 1, 1.0280000000000002e-06, 1.0280000000000002e-06, {('<frozen importlib._bootstrap>', 648, 'parent'): (1, 1, 1.0280000000000002e-06, 1.0280000000000002e-06)}), ('~', 0, '<built-in method builtins.getattr>'): (2, 2, 1.37e-06, 1.37e-06, {('numpy_support.py', 85, 'from_dtype'): (1, 1, 1.034e-06, 1.034e-06), ('misc.py', 63, 'unliteral'): (1, 1, 3.3600000000000004e-07, 3.3600000000000004e-07)}), ('~', 0, '<built-in method builtins.hasattr>'): (1, 1, 4.77e-07, 4.77e-07, {('misc.py', 63, 'unliteral'): (1, 1, 4.77e-07, 4.77e-07)}), ('~', 0, '<built-in method builtins.hash>'): (2, 4, 2.774e-06, 6.255e-06, {('abstract.py', 121, '__hash__'): (4, 2, 2.774e-06, 6.255e-06)}), ('~', 0, '<built-in method builtins.isinstance>'): (4, 4, 4.914000000000001e-06, 1.2491000000000001e-05, {('common.py', 49, '__init__'): (1, 1, 7.67e-07, 2.666e-06), ('npytypes.py', 444, '__init__'): (2, 2, 3.582e-06, 9.260000000000001e-06), ('typeof.py', 246, '_typeof_ndarray'): (1, 1, 5.65e-07, 5.65e-07)}), ('~', 0, '<built-in method _abc.get_cache_token>'): (1, 1, 5.82e-07, 5.82e-07, {('functools.py', 893, 'dispatch'): (1, 1, 5.82e-07, 5.82e-07)}), ('~', 0, '<built-in method _abc._abc_instancecheck>'): (3, 3, 4.903e-06, 4.903e-06, {('<frozen abc>', 117, '__instancecheck__'): (3, 3, 4.903e-06, 4.903e-06)})}

======================================================================
FAIL: test_call_event_chain (numba.tests.test_sys_monitoring.TestMonitoring.test_call_event_chain)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.14.3/x64/lib/python3.14/site-packages/numba/tests/test_sys_monitoring.py", line 201, in test_call_event_chain
    self.check_py_return_calls(cb)
    ~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^
  File "/opt/hostedtoolcache/Python/3.14.3/x64/lib/python3.14/site-packages/numba/tests/test_sys_monitoring.py", line 132, in check_py_return_calls
    self.assertEqual(mockcalls.call_count, 2)
    ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: 1 != 2

======================================================================
FAIL: test_monitoring_multiple_threads (numba.tests.test_sys_monitoring.TestMonitoring.test_monitoring_multiple_threads)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.14.3/x64/lib/python3.14/site-packages/numba/tests/test_sys_monitoring.py", line 797, in test_monitoring_multiple_threads
    assert_empty_queue(q1)
    ~~~~~~~~~~~~~~~~~~^^^^
  File "/opt/hostedtoolcache/Python/3.14.3/x64/lib/python3.14/site-packages/numba/tests/test_sys_monitoring.py", line 795, in assert_empty_queue
    self.fail("queue supposed to be empty")
    ~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: queue supposed to be empty

======================================================================
FAIL: test_multiple_tool_id (numba.tests.test_sys_monitoring.TestMonitoring.test_multiple_tool_id)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.14.3/x64/lib/python3.14/site-packages/numba/tests/test_sys_monitoring.py", line 446, in test_multiple_tool_id
    self.check_py_start_calls(cov_tool)
    ~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.14.3/x64/lib/python3.14/site-packages/numba/tests/test_sys_monitoring.py", line 107, in check_py_start_calls
    self.assertEqual(mockcalls.call_count, 2)
    ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: 1 != 2

======================================================================
FAIL: test_raising_callback_unwinds_from_jit_on_raising_path (numba.tests.test_sys_monitoring.TestMonitoring.test_raising_callback_unwinds_from_jit_on_raising_path)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.14.3/x64/lib/python3.14/site-packages/numba/tests/test_sys_monitoring.py", line 673, in test_raising_callback_unwinds_from_jit_on_raising_path
    self.assertEqual(callback.call_count, 3)
    ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: 2 != 3

======================================================================
FAIL: test_mutation_from_objmode (numba.tests.test_sys_monitoring.TestMonitoring.test_mutation_from_objmode)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.14.3/x64/lib/python3.14/site-packages/numba/tests/test_sys_monitoring.py", line 388, in test_mutation_from_objmode
    self.assertEqual(callback.call_count, 2)
    ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: 1 != 2

======================================================================
FAIL: test_raising_callback_unwinds_from_jit_on_success_path (numba.tests.test_sys_monitoring.TestMonitoring.test_raising_callback_unwinds_from_jit_on_success_path)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.14.3/x64/lib/python3.14/site-packages/numba/tests/test_sys_monitoring.py", line 632, in test_raising_callback_unwinds_from_jit_on_success_path
    callback.assert_called_once()
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~^^
  File "/opt/hostedtoolcache/Python/3.14.3/x64/lib/python3.14/unittest/mock.py", line 964, in assert_called_once
    raise AssertionError(msg)
AssertionError: Expected 'mock' to have been called once. Called 0 times.

======================================================================
FAIL: test_raising_under_monitoring (numba.tests.test_sys_monitoring.TestMonitoring.test_raising_under_monitoring)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.14.3/x64/lib/python3.14/site-packages/numba/tests/test_sys_monitoring.py", line 507, in test_raising_under_monitoring
    self.assertEqual(foo_call.args[0], foo.py_func.__code__)
    ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: <code object test_raising_under_monitoring at 0x561644[123 chars] 459> != <code object foo at 0x7f4509eb65b0, file "/opt/hostedt[97 chars] 470>

======================================================================
FAIL: test_return_event (numba.tests.test_sys_monitoring.TestMonitoring.test_return_event)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.14.3/x64/lib/python3.14/site-packages/numba/tests/test_sys_monitoring.py", line 193, in test_return_event
    self.check_py_return_calls(cb)
    ~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^
  File "/opt/hostedtoolcache/Python/3.14.3/x64/lib/python3.14/site-packages/numba/tests/test_sys_monitoring.py", line 132, in check_py_return_calls
    self.assertEqual(mockcalls.call_count, 2)
    ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: 1 != 2

======================================================================
FAIL: test_start_event (numba.tests.test_sys_monitoring.TestMonitoring.test_start_event)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.14.3/x64/lib/python3.14/site-packages/numba/tests/test_sys_monitoring.py", line 186, in test_start_event
    self.check_py_start_calls(cb)
    ~~~~~~~~~~~~~~~~~~~~~~~~~^^^^
  File "/opt/hostedtoolcache/Python/3.14.3/x64/lib/python3.14/site-packages/numba/tests/test_sys_monitoring.py", line 107, in check_py_start_calls
    self.assertEqual(mockcalls.call_count, 2)
    ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: 1 != 2

======================================================================
FAIL: test_stop_iteration_under_monitoring (numba.tests.test_sys_monitoring.TestMonitoring.test_stop_iteration_under_monitoring)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.14.3/x64/lib/python3.14/site-packages/numba/tests/test_sys_monitoring.py", line 576, in test_stop_iteration_under_monitoring
    self.assertEqual(raise_callback.call_count, 3)
    ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: 2 != 3


```

I can reproduce this with `docker` on osx.

Dockerfile:

```
FROM --platform=linux/amd64 python:3.14-slim

RUN apt-get update && apt-get install -y \
    build-essential \
    gcc \
    g++ \
    make \
    cmake \
    git \
    vim \
    && rm -rf /var/lib/apt/lists/*

# Install Numba and deps via pip (no conda on 3.14 yet)
RUN pip install --upgrade pip && \
    pip install \
        numpy \
        numba \
        pytest

WORKDIR /work

CMD ["/bin/bash"]
```

Build and launch commands:

```
docker build -t numba-py314 .

docker run -it --rm \
  --platform linux/amd64 \
  -v /path/to/numba:/work \
  numba-py314
```

Execute and reproduce:

```
root@6a0b59170e4c:/work# python -m numba.runtests numba.tests.test_sys_monitoring
F.....Traceback (most recent call last):
  File "/usr/local/lib/python3.14/site-packages/numba/tests/test_sys_monitoring.py", line 762, in t1_work
    self.check_py_start_calls_multithreads(cb)
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^
  File "/usr/local/lib/python3.14/site-packages/numba/tests/test_sys_monitoring.py", line 93, in check_py_start_calls_multithreads
    self.assertEqual(len(matched), 2) # two types of call
    ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.14/unittest/case.py", line 925, in assertEqual
    assertion_func(first, second, msg=msg)
    ~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.14/unittest/case.py", line 918, in _baseAssertEqual
    raise self.failureException(msg)
AssertionError: 1 != 2

FFF..FFEFF.FF.
======================================================================
ERROR: test_raising_callback_unwinds_from_jit_on_unwind_path (numba.tests.test_sys_monitoring.TestMonitoring.test_raising_callback_unwinds_from_jit_on_unwind_path)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/lib/python3.14/site-packages/numba/tests/test_sys_monitoring.py", line 723, in test_raising_callback_unwinds_from_jit_on_unwind_path
    raising()
    ~~~~~~~^^
  File "/usr/local/lib/python3.14/site-packages/numba/tests/test_sys_monitoring.py", line 714, in raising
    raise LocalException(msg_execution)
numba.tests.test_sys_monitoring.TestMonitoring.test_raising_callback_unwinds_from_jit_on_unwind_path.<locals>.LocalException: deliberately broken execution

======================================================================
FAIL: test_call_event_chain (numba.tests.test_sys_monitoring.TestMonitoring.test_call_event_chain)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/lib/python3.14/site-packages/numba/tests/test_sys_monitoring.py", line 201, in test_call_event_chain
    self.check_py_return_calls(cb)
    ~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^
  File "/usr/local/lib/python3.14/site-packages/numba/tests/test_sys_monitoring.py", line 132, in check_py_return_calls
    self.assertEqual(mockcalls.call_count, 2)
    ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: 1 != 2

======================================================================
FAIL: test_monitoring_multiple_threads (numba.tests.test_sys_monitoring.TestMonitoring.test_monitoring_multiple_threads)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/lib/python3.14/site-packages/numba/tests/test_sys_monitoring.py", line 797, in test_monitoring_multiple_threads
    assert_empty_queue(q1)
    ~~~~~~~~~~~~~~~~~~^^^^
  File "/usr/local/lib/python3.14/site-packages/numba/tests/test_sys_monitoring.py", line 795, in assert_empty_queue
    self.fail("queue supposed to be empty")
    ~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: queue supposed to be empty

======================================================================
FAIL: test_multiple_tool_id (numba.tests.test_sys_monitoring.TestMonitoring.test_multiple_tool_id)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/lib/python3.14/site-packages/numba/tests/test_sys_monitoring.py", line 446, in test_multiple_tool_id
    self.check_py_start_calls(cov_tool)
    ~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^
  File "/usr/local/lib/python3.14/site-packages/numba/tests/test_sys_monitoring.py", line 107, in check_py_start_calls
    self.assertEqual(mockcalls.call_count, 2)
    ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: 1 != 2

======================================================================
FAIL: test_mutation_from_objmode (numba.tests.test_sys_monitoring.TestMonitoring.test_mutation_from_objmode)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/lib/python3.14/site-packages/numba/tests/test_sys_monitoring.py", line 388, in test_mutation_from_objmode
    self.assertEqual(callback.call_count, 2)
    ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: 1 != 2

======================================================================
FAIL: test_raising_callback_unwinds_from_jit_on_raising_path (numba.tests.test_sys_monitoring.TestMonitoring.test_raising_callback_unwinds_from_jit_on_raising_path)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/lib/python3.14/site-packages/numba/tests/test_sys_monitoring.py", line 673, in test_raising_callback_unwinds_from_jit_on_raising_path
    self.assertEqual(callback.call_count, 3)
    ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: 2 != 3

======================================================================
FAIL: test_raising_callback_unwinds_from_jit_on_success_path (numba.tests.test_sys_monitoring.TestMonitoring.test_raising_callback_unwinds_from_jit_on_success_path)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/lib/python3.14/site-packages/numba/tests/test_sys_monitoring.py", line 632, in test_raising_callback_unwinds_from_jit_on_success_path
    callback.assert_called_once()
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~^^
  File "/usr/local/lib/python3.14/unittest/mock.py", line 965, in assert_called_once
    raise AssertionError(msg)
AssertionError: Expected 'mock' to have been called once. Called 0 times.

======================================================================
FAIL: test_raising_under_monitoring (numba.tests.test_sys_monitoring.TestMonitoring.test_raising_under_monitoring)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/lib/python3.14/site-packages/numba/tests/test_sys_monitoring.py", line 507, in test_raising_under_monitoring
    self.assertEqual(foo_call.args[0], foo.py_func.__code__)
    ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: <code object test_raising_under_monitoring at 0x555556[95 chars] 459> != <code object foo at 0x7fffccb74c60, file "/usr/local/l[69 chars] 470>

======================================================================
FAIL: test_return_event (numba.tests.test_sys_monitoring.TestMonitoring.test_return_event)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/lib/python3.14/site-packages/numba/tests/test_sys_monitoring.py", line 193, in test_return_event
    self.check_py_return_calls(cb)
    ~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^
  File "/usr/local/lib/python3.14/site-packages/numba/tests/test_sys_monitoring.py", line 132, in check_py_return_calls
    self.assertEqual(mockcalls.call_count, 2)
    ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: 1 != 2

======================================================================
FAIL: test_start_event (numba.tests.test_sys_monitoring.TestMonitoring.test_start_event)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/lib/python3.14/site-packages/numba/tests/test_sys_monitoring.py", line 186, in test_start_event
    self.check_py_start_calls(cb)
    ~~~~~~~~~~~~~~~~~~~~~~~~~^^^^
  File "/usr/local/lib/python3.14/site-packages/numba/tests/test_sys_monitoring.py", line 107, in check_py_start_calls
    self.assertEqual(mockcalls.call_count, 2)
    ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: 1 != 2

======================================================================
FAIL: test_stop_iteration_under_monitoring (numba.tests.test_sys_monitoring.TestMonitoring.test_stop_iteration_under_monitoring)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/lib/python3.14/site-packages/numba/tests/test_sys_monitoring.py", line 576, in test_stop_iteration_under_monitoring
    self.assertEqual(raise_callback.call_count, 3)
    ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: 2 != 3

----------------------------------------------------------------------
Ran 20 tests in 10.048s

FAILED (failures=10, errors=1)
```

--- a/docs/source/developer/sys_monitoring.rst
+++ b/docs/source/developer/sys_monitoring.rst
@@ -7,11 +7,6 @@ Notes on ``sys.monitoring``
           versions of Python may behave differently. It is however hoped that
           most of the concepts herein will remain relevant.
 
-.. note:: From Python 3.14.4 onward the dispatcher behaviour described below is
-          disabled (`#10538 <https://github.com/numba/numba/issues/10538>`_).
-          ``NUMBA_ENABLE_SYS_MONITORING`` has no effect on those versions and a
-          ``UserWarning`` is emitted if this variable is set to a non-zero value.
-
 Python 3.12 introduced a new monitoring system under ``sys.monitoring``. This
 system lets users monitor a selection of events that may be interesting for e.g.
 performance profiling or debugging purposes. Event monitoring is set "per tool",

--- a/docs/source/reference/envvars.rst
+++ b/docs/source/reference/envvars.rst
@@ -194,12 +194,6 @@ These variables influence what is printed out during compilation of
 
    Only available for Python 3.12 and above. Otherwise, it has no effect.
 
-   It also has no effect on Python 3.14.4 and newer: dispatcher support for
-   ``sys.monitoring`` is disabled, see this issue for more details
-   (`#10538 <https://github.com/numba/numba/issues/10538>`_). A
-   ``UserWarning`` is emitted if this variable is set to a
-   non-zero value.
-
 .. envvar:: NUMBA_ENABLE_PROFILING
 
    Enables JIT events of LLVM in order to support profiling of jitted functions.

--- a/docs/upcoming_changes/10547.change.rst
+++ b/docs/upcoming_changes/10547.change.rst
@@ -1,7 +1,0 @@
-Disable JIT ``sys.monitoring`` on Python 3.14.4 and later
----------------------------------------------------------
-
-Python 3.14.4 changed interpreter internals Numba relied on for
-``sys.monitoring`` (`#10538 <https://github.com/numba/numba/issues/10538>`_).
-The feature is disabled from 3.14.4 onward; ``NUMBA_ENABLE_SYS_MONITORING`` has no
-effect and a ``UserWarning`` will be emitted if this variable is set to a non-zero value.

--- a/docs/upcoming_changes/10578.change.rst
+++ b/docs/upcoming_changes/10578.change.rst
@@ -1,0 +1,12 @@
+Use public ``PyMonitoring_*`` C API for ``sys.monitoring`` on Python 3.13+
+--------------------------------------------------------------------------
+
+Numba's dispatcher integration with ``sys.monitoring`` on Python 3.13 and
+later now goes through the public ``PyMonitoring_*`` C API instead of
+CPython internal headers. This restores ``sys.monitoring`` support on
+Python 3.14.4 and newer (`#10538
+<https://github.com/numba/numba/issues/10538>`_), which had been disabled
+due to ABI changes in interpreter internals Numba previously relied on.
+``NUMBA_ENABLE_SYS_MONITORING`` again takes effect on those versions and
+the associated ``UserWarning`` is no longer emitted. The Python 3.12 code
+path is unchanged.

--- a/numba/_dispatcher.cpp
+++ b/numba/_dispatcher.cpp
@@ -812,43 +812,6 @@ static inline int msb(uint8_t bits) {
     return MOST_SIG_BIT[bits];
 }
 
-/*
- * On Python 3.14.4+, skip JIT sys.monitoring (#10538).
- * https://github.com/numba/numba/issues/10538
- */
-static int
-jit_sysmon_supported(void)
-{
-    // support_flag states: -1 unknown, 0 unsupported (warn once), 1 supported
-    static int support_flag = -1;
-    unsigned long v;
-    int major, minor, micro;
-
-    if (support_flag != -1) {
-        return support_flag;
-    }
-
-    v = Py_Version;
-    major = (int)((v >> 24) & 0xFF);
-    minor = (int)((v >> 16) & 0xFF);
-    micro = (int)((v >> 8) & 0xFF);
-
-    // Exclude anything beyond 3.14.3.
-    support_flag = !(major == 3 && (minor > 14 || (minor == 14 && micro > 3)));
-
-    if (!support_flag) {
-        PyErr_WarnEx(PyExc_UserWarning,
-            "Numba: JIT sys.monitoring integration is disabled on "
-            "Python 3.14.4+ (https://github.com/numba/numba/"
-            "issues/10538). "
-            "Unset NUMBA_ENABLE_SYS_MONITORING to suppress this warning.",
-            1
-        );
-    }
-
-    return support_flag;
-}
-
 
 static int invoke_monitoring(PyThreadState * tstate, int event, Dispatcher *self, PyObject* retval)
 {
@@ -939,10 +902,6 @@ static int invoke_monitoring(PyThreadState * tstate, int event, Dispatcher *self
     // https://github.com/python/cpython/blob/0ab2384c5f56625e99bb35417cadddfe24d347e1/Python/instrumentation.c#L945-L1008
     // https://github.com/python/cpython/blob/0ab2384c5f56625e99bb35417cadddfe24d347e1/Python/instrumentation.c#L1010-L1026
     // https://github.com/python/cpython/blob/0ab2384c5f56625e99bb35417cadddfe24d347e1/Python/instrumentation.c#L839-L861
-
-    if (!jit_sysmon_supported()) {
-        return 0;
-    }
 
     // TODO: check this, call_instrumentation_vector has this at the top.
     if (tstate->tracing){

--- a/numba/_dispatcher.cpp
+++ b/numba/_dispatcher.cpp
@@ -50,7 +50,6 @@
 #endif
 #include "internal/pycore_interp.h"
 #include "internal/pycore_pyerrors.h"
-#include "internal/pycore_instruments.h"
 #include "internal/pycore_call.h"
 #include "cpython/code.h"
 
@@ -791,10 +790,13 @@ call_cfunc(Dispatcher *self, PyObject *cfunc, PyObject *args, PyObject *kws, PyO
     }
 }
 
-#elif (PY_MAJOR_VERSION >= 3) && ((PY_MINOR_VERSION == 12) || (PY_MINOR_VERSION == 13) || (PY_MINOR_VERSION == 14))
+#elif (PY_MAJOR_VERSION >= 3) && (PY_MINOR_VERSION == 12)
+// Python 3.12 only: use invoke_monitoring() that reads
+// CPython internal struct fields directly.
+// NOTE: Do NOT extend this block to 3.13+. Use the public
+// PyMonitoring_Fire*Event API below instead.
 
-// Python 3.12 has a completely new approach to tracing and profiling due to
-// the new `sys.monitoring` system.
+#include "internal/pycore_instruments.h"
 
 // From: https://github.com/python/cpython/blob/0ab2384c5f56625e99bb35417cadddfe24d347e1/Python/instrumentation.c#L863-L868
 
@@ -1122,6 +1124,107 @@ call_cfunc(Dispatcher *self, PyObject *cfunc, PyObject *args, PyObject *kws, PyO
     if(enabled_sysmon && invoke_monitoring_PY_RETURN(tstate, self, pyresult) != 0){
         return NULL;
     }
+    return pyresult;
+}
+
+#elif (PY_MAJOR_VERSION >= 3) && (PY_MINOR_VERSION >= 13)
+
+// Monitoring event indices into local mon_states[].
+enum {
+    NUMBA_MON_PY_START   = 0,
+    NUMBA_MON_PY_RETURN  = 1,
+    NUMBA_MON_RAISE      = 2,
+    NUMBA_MON_PY_UNWIND  = 3,
+    NUMBA_MON_COUNT      = 4,
+};
+
+// The event types we care about, in the same order as the enum above.
+static const uint8_t numba_mon_event_types[NUMBA_MON_COUNT] = {
+    PY_MONITORING_EVENT_PY_START,
+    PY_MONITORING_EVENT_PY_RETURN,
+    PY_MONITORING_EVENT_RAISE,
+    PY_MONITORING_EVENT_PY_UNWIND,
+};
+
+// Refresh monitoring state via the public API. This reads the interpreter's
+// monitoring configuration through libpython at runtime.
+static int refresh_monitoring_scope(PyMonitoringState mon_states[],
+                                    uint64_t *mon_version) {
+    return PyMonitoring_EnterScope(
+        mon_states,
+        mon_version,
+        numba_mon_event_types,
+        NUMBA_MON_COUNT);
+}
+
+/* forward declaration */
+static bool is_sysmon_enabled(Dispatcher *self);
+
+static PyObject *
+call_cfunc(Dispatcher *self, PyObject *cfunc, PyObject *args, PyObject *kws, PyObject *locals)
+{
+    PyCFunctionWithKeywords fn = NULL;
+    PyObject *pyresult = NULL;
+    PyObject *codelike = NULL;
+    const bool enabled_sysmon = is_sysmon_enabled(self);
+    int in_scope = 0;
+    PyMonitoringState mon_states[NUMBA_MON_COUNT];
+    uint64_t mon_version = 0;
+
+    assert(PyCFunction_Check(cfunc));
+    assert(PyCFunction_GET_FLAGS(cfunc) == (METH_VARARGS | METH_KEYWORDS));
+    fn = (PyCFunctionWithKeywords) PyCFunction_GET_FUNCTION(cfunc);
+
+    if (enabled_sysmon) {
+        if (refresh_monitoring_scope(mon_states, &mon_version) < 0)
+            return NULL;
+        in_scope = 1;
+
+        codelike = PyObject_GetAttrString((PyObject*)self, "__code__");
+        if (!codelike)
+            goto exit_scope;
+
+        if (PyMonitoring_FirePyStartEvent(
+                &mon_states[NUMBA_MON_PY_START], codelike, 0) < 0)
+            goto exit_scope;
+    }
+
+    pyresult = fn(PyCFunction_GET_SELF(cfunc), args, kws);
+
+    if (enabled_sysmon && pyresult == NULL) {
+        // Exception path. Refresh scope — events may have been toggled
+        // during objmode. Leave the exception on the error indicator;
+        // Fire*Event reads it from tstate->current_exception internally.
+        if (refresh_monitoring_scope(mon_states, &mon_version) < 0)
+            goto exit_scope;
+        if (PyErr_Occurred()) {
+            if (PyMonitoring_FireRaiseEvent(
+                    &mon_states[NUMBA_MON_RAISE], codelike, 0) < 0)
+                goto exit_scope;
+            if (PyMonitoring_FirePyUnwindEvent(
+                    &mon_states[NUMBA_MON_PY_UNWIND], codelike, 0) < 0)
+                goto exit_scope;
+        }
+        goto exit_scope;
+    }
+
+    if (enabled_sysmon) {
+        // Normal return. Refresh scope — events may have been toggled
+        // during objmode.
+        if (refresh_monitoring_scope(mon_states, &mon_version) < 0)
+            goto exit_scope;
+        if (PyMonitoring_FirePyReturnEvent(
+                &mon_states[NUMBA_MON_PY_RETURN], codelike, 0,
+                pyresult) < 0) {
+            Py_CLEAR(pyresult);
+            goto exit_scope;
+        }
+    }
+
+exit_scope:
+    if (in_scope)
+        PyMonitoring_ExitScope();
+    Py_XDECREF(codelike);
     return pyresult;
 }
 #else

--- a/numba/_dispatcher.cpp
+++ b/numba/_dispatcher.cpp
@@ -1168,7 +1168,9 @@ call_cfunc(Dispatcher *self, PyObject *cfunc, PyObject *args, PyObject *kws, PyO
     PyObject *codelike = NULL;
     const bool enabled_sysmon = is_sysmon_enabled(self);
     int in_scope = 0;
-    PyMonitoringState mon_states[NUMBA_MON_COUNT];
+    // initialize mon_states as inactive; refresh_monitoring_scope skips
+    // writing them when the monitoring version is unchanged.
+    PyMonitoringState mon_states[NUMBA_MON_COUNT] = {};
     uint64_t mon_version = 0;
 
     assert(PyCFunction_Check(cfunc));

--- a/numba/_dispatcher.cpp
+++ b/numba/_dispatcher.cpp
@@ -1178,18 +1178,18 @@ call_cfunc(Dispatcher *self, PyObject *cfunc, PyObject *args, PyObject *kws, PyO
     fn = (PyCFunctionWithKeywords) PyCFunction_GET_FUNCTION(cfunc);
 
     if (enabled_sysmon) {
-        if (refresh_monitoring_scope(mon_states, &mon_version) < 0){
+        if (refresh_monitoring_scope(mon_states, &mon_version) < 0) {
             return NULL;
         }
         in_scope = 1;
 
         codelike = PyObject_GetAttrString((PyObject*)self, "__code__");
-        if (!codelike){
+        if (!codelike) {
             goto exit_scope;
         }
 
         if (PyMonitoring_FirePyStartEvent(
-                &mon_states[NUMBA_MON_PY_START], codelike, 0) < 0){
+                &mon_states[NUMBA_MON_PY_START], codelike, 0) < 0) {
             goto exit_scope;
         }
     }
@@ -1200,16 +1200,16 @@ call_cfunc(Dispatcher *self, PyObject *cfunc, PyObject *args, PyObject *kws, PyO
         // Exception path. Refresh scope — events may have been toggled
         // during objmode. Leave the exception on the error indicator;
         // Fire*Event reads it from tstate->current_exception internally.
-        if (refresh_monitoring_scope(mon_states, &mon_version) < 0){
+        if (refresh_monitoring_scope(mon_states, &mon_version) < 0) {
             goto exit_scope;
         }
         if (PyErr_Occurred()) {
             if (PyMonitoring_FireRaiseEvent(
-                    &mon_states[NUMBA_MON_RAISE], codelike, 0) < 0){
+                    &mon_states[NUMBA_MON_RAISE], codelike, 0) < 0) {
                 goto exit_scope;
             }
             if (PyMonitoring_FirePyUnwindEvent(
-                    &mon_states[NUMBA_MON_PY_UNWIND], codelike, 0) < 0){
+                    &mon_states[NUMBA_MON_PY_UNWIND], codelike, 0) < 0) {
                 goto exit_scope;
             }
         }
@@ -1219,7 +1219,7 @@ call_cfunc(Dispatcher *self, PyObject *cfunc, PyObject *args, PyObject *kws, PyO
     if (enabled_sysmon) {
         // Normal return. Refresh scope — events may have been toggled
         // during objmode.
-        if (refresh_monitoring_scope(mon_states, &mon_version) < 0){
+        if (refresh_monitoring_scope(mon_states, &mon_version) < 0) {
             Py_CLEAR(pyresult);
             goto exit_scope;
         }
@@ -1232,7 +1232,7 @@ call_cfunc(Dispatcher *self, PyObject *cfunc, PyObject *args, PyObject *kws, PyO
     }
 
 exit_scope:
-    if (in_scope){
+    if (in_scope) {
         PyMonitoring_ExitScope();
     }
     Py_XDECREF(codelike);

--- a/numba/_dispatcher.cpp
+++ b/numba/_dispatcher.cpp
@@ -1178,17 +1178,20 @@ call_cfunc(Dispatcher *self, PyObject *cfunc, PyObject *args, PyObject *kws, PyO
     fn = (PyCFunctionWithKeywords) PyCFunction_GET_FUNCTION(cfunc);
 
     if (enabled_sysmon) {
-        if (refresh_monitoring_scope(mon_states, &mon_version) < 0)
+        if (refresh_monitoring_scope(mon_states, &mon_version) < 0){
             return NULL;
+        }
         in_scope = 1;
 
         codelike = PyObject_GetAttrString((PyObject*)self, "__code__");
-        if (!codelike)
+        if (!codelike){
             goto exit_scope;
+        }
 
         if (PyMonitoring_FirePyStartEvent(
-                &mon_states[NUMBA_MON_PY_START], codelike, 0) < 0)
+                &mon_states[NUMBA_MON_PY_START], codelike, 0) < 0){
             goto exit_scope;
+        }
     }
 
     pyresult = fn(PyCFunction_GET_SELF(cfunc), args, kws);
@@ -1197,15 +1200,18 @@ call_cfunc(Dispatcher *self, PyObject *cfunc, PyObject *args, PyObject *kws, PyO
         // Exception path. Refresh scope — events may have been toggled
         // during objmode. Leave the exception on the error indicator;
         // Fire*Event reads it from tstate->current_exception internally.
-        if (refresh_monitoring_scope(mon_states, &mon_version) < 0)
+        if (refresh_monitoring_scope(mon_states, &mon_version) < 0){
             goto exit_scope;
+        }
         if (PyErr_Occurred()) {
             if (PyMonitoring_FireRaiseEvent(
-                    &mon_states[NUMBA_MON_RAISE], codelike, 0) < 0)
+                    &mon_states[NUMBA_MON_RAISE], codelike, 0) < 0){
                 goto exit_scope;
+            }
             if (PyMonitoring_FirePyUnwindEvent(
-                    &mon_states[NUMBA_MON_PY_UNWIND], codelike, 0) < 0)
+                    &mon_states[NUMBA_MON_PY_UNWIND], codelike, 0) < 0){
                 goto exit_scope;
+            }
         }
         goto exit_scope;
     }
@@ -1213,8 +1219,9 @@ call_cfunc(Dispatcher *self, PyObject *cfunc, PyObject *args, PyObject *kws, PyO
     if (enabled_sysmon) {
         // Normal return. Refresh scope — events may have been toggled
         // during objmode.
-        if (refresh_monitoring_scope(mon_states, &mon_version) < 0)
+        if (refresh_monitoring_scope(mon_states, &mon_version) < 0){
             goto exit_scope;
+        }
         if (PyMonitoring_FirePyReturnEvent(
                 &mon_states[NUMBA_MON_PY_RETURN], codelike, 0,
                 pyresult) < 0) {
@@ -1224,8 +1231,9 @@ call_cfunc(Dispatcher *self, PyObject *cfunc, PyObject *args, PyObject *kws, PyO
     }
 
 exit_scope:
-    if (in_scope)
+    if (in_scope){
         PyMonitoring_ExitScope();
+    }
     Py_XDECREF(codelike);
     return pyresult;
 }

--- a/numba/_dispatcher.cpp
+++ b/numba/_dispatcher.cpp
@@ -1220,6 +1220,7 @@ call_cfunc(Dispatcher *self, PyObject *cfunc, PyObject *args, PyObject *kws, PyO
         // Normal return. Refresh scope — events may have been toggled
         // during objmode.
         if (refresh_monitoring_scope(mon_states, &mon_version) < 0){
+            Py_CLEAR(pyresult);
             goto exit_scope;
         }
         if (PyMonitoring_FirePyReturnEvent(

--- a/numba/_dispatcher.cpp
+++ b/numba/_dispatcher.cpp
@@ -1181,7 +1181,7 @@ call_cfunc(Dispatcher *self, PyObject *cfunc, PyObject *args, PyObject *kws, PyO
         return fn(PyCFunction_GET_SELF(cfunc), args, kws);
     }
 
-    if (refresh_monitoring_scope(mon_states, &mon_version) < 0) {
+    if (refresh_monitoring_scope(mon_states, &mon_version) != 0) {
         return NULL;
     }
     in_scope = 1;
@@ -1192,7 +1192,7 @@ call_cfunc(Dispatcher *self, PyObject *cfunc, PyObject *args, PyObject *kws, PyO
     }
 
     if (PyMonitoring_FirePyStartEvent(
-                &mon_states[NUMBA_MON_PY_START], codelike, 0) < 0) {
+                &mon_states[NUMBA_MON_PY_START], codelike, 0) != 0) {
         goto exit_scope;
     }
 
@@ -1202,16 +1202,16 @@ call_cfunc(Dispatcher *self, PyObject *cfunc, PyObject *args, PyObject *kws, PyO
         // Exception path. Refresh scope — events may have been toggled
         // during objmode. Leave the exception on the error indicator;
         // Fire*Event reads it from tstate->current_exception internally.
-        if (refresh_monitoring_scope(mon_states, &mon_version) < 0) {
+        if (refresh_monitoring_scope(mon_states, &mon_version) != 0) {
             goto exit_scope;
         }
         if (PyErr_Occurred()) {
             if (PyMonitoring_FireRaiseEvent(
-                    &mon_states[NUMBA_MON_RAISE], codelike, 0) < 0) {
+                    &mon_states[NUMBA_MON_RAISE], codelike, 0) != 0) {
                 goto exit_scope;
             }
             if (PyMonitoring_FirePyUnwindEvent(
-                    &mon_states[NUMBA_MON_PY_UNWIND], codelike, 0) < 0) {
+                    &mon_states[NUMBA_MON_PY_UNWIND], codelike, 0) != 0) {
                 goto exit_scope;
             }
         }
@@ -1220,13 +1220,13 @@ call_cfunc(Dispatcher *self, PyObject *cfunc, PyObject *args, PyObject *kws, PyO
 
     // Normal return. Refresh scope — events may have been toggled
     // during objmode.
-    if (refresh_monitoring_scope(mon_states, &mon_version) < 0) {
+    if (refresh_monitoring_scope(mon_states, &mon_version) != 0) {
         Py_CLEAR(pyresult);
         goto exit_scope;
     }
     if (PyMonitoring_FirePyReturnEvent(
             &mon_states[NUMBA_MON_PY_RETURN], codelike, 0,
-            pyresult) < 0) {
+            pyresult) != 0) {
         Py_CLEAR(pyresult);
         goto exit_scope;
     }

--- a/numba/_dispatcher.cpp
+++ b/numba/_dispatcher.cpp
@@ -1177,26 +1177,28 @@ call_cfunc(Dispatcher *self, PyObject *cfunc, PyObject *args, PyObject *kws, PyO
     assert(PyCFunction_GET_FLAGS(cfunc) == (METH_VARARGS | METH_KEYWORDS));
     fn = (PyCFunctionWithKeywords) PyCFunction_GET_FUNCTION(cfunc);
 
-    if (enabled_sysmon) {
-        if (refresh_monitoring_scope(mon_states, &mon_version) < 0) {
-            return NULL;
-        }
-        in_scope = 1;
+    if (!enabled_sysmon) {
+        return fn(PyCFunction_GET_SELF(cfunc), args, kws);
+    }
 
-        codelike = PyObject_GetAttrString((PyObject*)self, "__code__");
-        if (!codelike) {
-            goto exit_scope;
-        }
+    if (refresh_monitoring_scope(mon_states, &mon_version) < 0) {
+        return NULL;
+    }
+    in_scope = 1;
 
-        if (PyMonitoring_FirePyStartEvent(
+    codelike = PyObject_GetAttrString((PyObject*)self, "__code__");
+    if (codelike == NULL) {
+        goto exit_scope;
+    }
+
+    if (PyMonitoring_FirePyStartEvent(
                 &mon_states[NUMBA_MON_PY_START], codelike, 0) < 0) {
-            goto exit_scope;
-        }
+        goto exit_scope;
     }
 
     pyresult = fn(PyCFunction_GET_SELF(cfunc), args, kws);
 
-    if (enabled_sysmon && pyresult == NULL) {
+    if (pyresult == NULL) {
         // Exception path. Refresh scope — events may have been toggled
         // during objmode. Leave the exception on the error indicator;
         // Fire*Event reads it from tstate->current_exception internally.
@@ -1216,19 +1218,17 @@ call_cfunc(Dispatcher *self, PyObject *cfunc, PyObject *args, PyObject *kws, PyO
         goto exit_scope;
     }
 
-    if (enabled_sysmon) {
-        // Normal return. Refresh scope — events may have been toggled
-        // during objmode.
-        if (refresh_monitoring_scope(mon_states, &mon_version) < 0) {
-            Py_CLEAR(pyresult);
-            goto exit_scope;
-        }
-        if (PyMonitoring_FirePyReturnEvent(
-                &mon_states[NUMBA_MON_PY_RETURN], codelike, 0,
-                pyresult) < 0) {
-            Py_CLEAR(pyresult);
-            goto exit_scope;
-        }
+    // Normal return. Refresh scope — events may have been toggled
+    // during objmode.
+    if (refresh_monitoring_scope(mon_states, &mon_version) < 0) {
+        Py_CLEAR(pyresult);
+        goto exit_scope;
+    }
+    if (PyMonitoring_FirePyReturnEvent(
+            &mon_states[NUMBA_MON_PY_RETURN], codelike, 0,
+            pyresult) < 0) {
+        Py_CLEAR(pyresult);
+        goto exit_scope;
     }
 
 exit_scope:

--- a/numba/tests/support.py
+++ b/numba/tests/support.py
@@ -143,11 +143,6 @@ skip_if_freethreading = unittest.skipIf(_free_threading,
                                          "free-threading is enabled."))
 
 
-skip_if_sysmon_unsupported = unittest.skipIf(
-    sys.version_info[:3] >= (3, 14, 4),
-    "JIT sys.monitoring is not supported on Python 3.14.4+")
-
-
 def expected_failure_py311(fn):
     if utils.PYVERSION == (3, 11):
         return unittest.expectedFailure(fn)

--- a/numba/tests/test_profiler.py
+++ b/numba/tests/test_profiler.py
@@ -7,7 +7,7 @@ import sys
 import numpy as np
 
 from numba import jit, types
-from numba.tests.support import needs_blas, skip_if_sysmon_unsupported
+from numba.tests.support import needs_blas
 import unittest
 
 
@@ -86,12 +86,10 @@ class TestProfiler(unittest.TestCase):
         if caller is not cfunc:
             check_stats_for_key(stats, caller.__code__, n_calls)
 
-    @skip_if_sysmon_unsupported
     def test_profiler(self):
         dot, _ = generate_standard_dot_case()
         self.check_profiler_dot(dot, dot)
 
-    @skip_if_sysmon_unsupported
     def test_profiler_for_raising_function(self):
         raising_dot, call_raising_dot = generate_raising_dot_case()
         self.check_profiler_dot(call_raising_dot, raising_dot)

--- a/numba/tests/test_sys_monitoring.py
+++ b/numba/tests/test_sys_monitoring.py
@@ -4,10 +4,9 @@ import sys
 import threading
 import traceback
 import unittest
-import warnings
 from collections import Counter
 from unittest.mock import Mock, call
-from numba.tests.support import TestCase, skip_if_sysmon_unsupported
+from numba.tests.support import TestCase
 from numba import jit, objmode
 from numba.core.utils import PYVERSION
 from numba.core.serialize import _numba_unpickle
@@ -179,7 +178,6 @@ class TestMonitoring(TestCase):
             sys.monitoring.free_tool_id(_tool_id)
         return callbacks
 
-    @skip_if_sysmon_unsupported
     def test_start_event(self):
         # test event PY_START
         cb = self.run_with_events(self.call_foo, (self.arg,), (PY_START,))
@@ -187,7 +185,6 @@ class TestMonitoring(TestCase):
         self.assertEqual(len(cb), 1)
         self.check_py_start_calls(cb)
 
-    @skip_if_sysmon_unsupported
     def test_return_event(self):
         # test event PY_RETURN
         cb = self.run_with_events(self.call_foo, (self.arg,), (PY_RETURN,))
@@ -195,7 +192,6 @@ class TestMonitoring(TestCase):
         self.assertEqual(len(cb), 1)
         self.check_py_return_calls(cb)
 
-    @skip_if_sysmon_unsupported
     def test_call_event_chain(self):
         # test event PY_START and PY_RETURN monitored at the same time
         cb = self.run_with_events(self.call_foo, (self.arg,),
@@ -317,7 +313,6 @@ class TestMonitoring(TestCase):
             sys.monitoring.set_events(tool_id, NO_EVENTS)
             sys.monitoring.free_tool_id(tool_id)
 
-    @skip_if_sysmon_unsupported
     def test_disable_from_callback(self):
         # Event callbacks can disable a _local_ event at a specific location to
         # prevent it triggering in the future by returning
@@ -354,7 +349,6 @@ class TestMonitoring(TestCase):
             sys.monitoring.set_events(tool_id, NO_EVENTS)
             sys.monitoring.free_tool_id(tool_id)
 
-    @skip_if_sysmon_unsupported
     def test_mutation_from_objmode(self):
         try:
             # Check that it's possible to enable an event (mutate the event
@@ -397,7 +391,6 @@ class TestMonitoring(TestCase):
             sys.monitoring.register_callback(tool_id, event, None)
             sys.monitoring.free_tool_id(tool_id)
 
-    @skip_if_sysmon_unsupported
     def test_multiple_tool_id(self):
         # Check that multiple tools will work across different combinations of
         # events that Numba dispatcher supports, namely:
@@ -463,7 +456,6 @@ class TestMonitoring(TestCase):
         self.check_py_start_calls(opt_tool)
         self.check_py_return_calls(opt_tool)
 
-    @skip_if_sysmon_unsupported
     def test_raising_under_monitoring(self):
         # Check that Numba can raise an exception whilst monitoring is running
         # and that 1. `RAISE` is issued 2. `PY_UNWIND` is issued, 3. that
@@ -531,7 +523,6 @@ class TestMonitoring(TestCase):
 
         self.assertIn(msg, str(store_raised))
 
-    @skip_if_sysmon_unsupported
     def test_stop_iteration_under_monitoring(self):
         # Check that Numba can raise an StopIteration exception whilst
         # monitoring is running and that:
@@ -614,7 +605,6 @@ class TestMonitoring(TestCase):
 
         self.assertIn(msg, str(store_raised))
 
-    @skip_if_sysmon_unsupported
     def test_raising_callback_unwinds_from_jit_on_success_path(self):
         # An event callback can legitimately raise an exception, this test
         # makes sure Numba's dispatcher handles it ok on the "successful path",
@@ -642,7 +632,6 @@ class TestMonitoring(TestCase):
         callback.assert_called_once()
         self.assertIn(msg, str(store_raised))
 
-    @skip_if_sysmon_unsupported
     def test_raising_callback_unwinds_from_jit_on_raising_path(self):
         # An event callback can legitimately raise an exception, this test
         # makes sure Numba's dispatcher handles it ok on the
@@ -702,7 +691,6 @@ class TestMonitoring(TestCase):
         # check the stored exception is the expected exception
         self.assertIs(store_raised, callback.side_effect)
 
-    @skip_if_sysmon_unsupported
     def test_raising_callback_unwinds_from_jit_on_unwind_path(self):
         # An event callback can legitimately raise an exception, this test
         # makes sure Numba's dispatcher handles it ok on the
@@ -751,7 +739,6 @@ class TestMonitoring(TestCase):
         # check the stored_raise
         self.assertIs(store_raised, callback.side_effect)
 
-    @skip_if_sysmon_unsupported
     def test_monitoring_multiple_threads(self):
         # Two threads, different tools and events registered on each thread.
         # Each test creates a global event capturing. The threads use barriers
@@ -853,31 +840,6 @@ class TestMonitoringEnvVarControl(TestCase):
             return x + 1
 
         self.assertTrue(foo._enable_sysmon)
-
-    @unittest.skipUnless(
-        sys.version_info[:3] >= (3, 14, 4),
-        "needs Python 3.14.4+",
-    )
-    @TestCase.run_test_in_subprocess(
-        envvars={"NUMBA_ENABLE_SYS_MONITORING": '1'})
-    def test_userwarning_when_jit_sysmon_unavailable(self):
-        # C dispatcher warns once when hooks are off on 3.14.4+ (#10538).
-        from numba import jit
-
-        with warnings.catch_warnings(record=True) as w:
-            warnings.simplefilter("always", UserWarning)
-
-            @jit(nopython=True)
-            def foo(x):
-                return x + 1
-
-            self.assertTrue(foo._enable_sysmon)
-            self.assertEqual(foo(1), 2)
-            self.assertEqual(foo(2), 3)
-
-        self.assertEqual(len(w), 1)
-        msg = str(w[0].message)
-        self.assertIn("sys.monitoring", msg)
 
 
 if __name__ == '__main__':

